### ISSUE TITLE
PROXYFS has a problem with reporting file sizes and SEEK_END

### ISSUE THREAD

**Issue by vadimkantorov** [emscripten-core/emscripten#12367]
This code from dvidpfmx reports 0 file size when using PROXYFS and correct 1924 size when using MEMFS:
```
off_t xfile_size (FILE *file, const char *name)
{
  off_t size;
  xseek_end (file, name);
  size = xtell_position (file, name);
  rewind (file);
  return size;
}
```

*UPD* old content is below:

[Here](https://github.com/emscripten-core/emscripten/blob/7c3ced64d51be643f7a12bb34e8b875155d49dc7/src/library_proxyfs.js#L195) it supports only SEEK_CUR and SEEK_END.

SEEK_SET is also useful and is used e.g. by xdvpdfmx:
https://github.com/TeX-Live/texlive-source/blob/ec443c5d5ef89295f03bbd54d5f1f77ebfd7e6b4/texk/dvipdfm-x/dvi.c#L365
https://github.com/TeX-Live/texlive-source/blob/ec443c5d5ef89295f03bbd54d5f1f77ebfd7e6b4/texk/dvipdfm-x/mfileio.h#L56

**Comment by sbc100** [comment]
I guess this is not very surprising, proxyfs seems largely undocumented/untested/unused.  

Would you like to help out improving it?

Is this even something we want to encorage or move forward with?  @azakai @juj?

**Comment by vadimkantorov** [comment]
It is very useful for having several programs/modules work with the same FS. Initializing FS can be costly if there's a LOOT of files (even if virtual files), just because running maany javascript calls can be slow. It is also a simple example of creating a user FS.

I don't have resources right now for fast-timeline-ish contributions :(, it would take a lot of time on my side (to learn Emscripten's CI, testing infra etc). Yet I'll try to see how such SEEK op should be implemented

**Comment by sbc100** [comment]
I don't mean to sounds like your use case is not important.   We just have to consider the cost of maintaining each feature against its usage given our limited resources.

For example, if you are the only user of this feature it might make sense to maintain the `library_proxyfs.js` outside of emscripten.  This is possible in this case because of the pluggable nature of the filesystem API.

I'm not saying we have to go with direction, just putting some options out there.  There could well be other users of this feature that I'm not aware of.

In any case we would welcome contributions to the existing code!

**Comment by vadimkantorov** [comment]
Regardless of PROXYFS, what is needed is more information / documentation about creating custom FS's and the respecive limitations. E.g. I may want to add a caching Web-fetching FS (think of mounting online TexLive database). Examples are useful, maybe PROXYFS could be turned into such an example.

In theory, could I just copy over PROXYFS into my own code? and just remove the prefix
```
mergeInto(LibraryManager.library, {
  $PROXYFS__deps: ['$FS', '$PATH', '$ERRNO_CODES'],
  $PROXYFS: 
```

**Comment by vadimkantorov** [comment]
The problem is in negative offsets

**Comment by vadimkantorov** [comment]
The problem seems to be in `tell(...)` call

**Comment by vadimkantorov** [comment]
Somehow SEEK_END call does not modify stream.position. I'm not sure how stream's cursor gets modified. Not sure how it works for MEMFS

**Comment by vadimkantorov** [comment]
So after a proxied lseek(SEEK_END) call stream.position still returns 0 on the next call

**Comment by vadimkantorov** [comment]
How is stream.position supposed to move? MEMFS also does not advance it: 
https://github.com/emscripten-core/emscripten/blob/7c3ced64d51be643f7a12bb34e8b875155d49dc7/src/library_memfs.js#L339

The pattern in dvipdfmx for learning the file size is to SEEK_END and then to retrieve SEEK_CUR. This currently retrieves zero

**Comment by vadimkantorov** [comment]
@sbc100 Okay, I figured it out!

PROXYFS's handles SEEK_END by fstat call, but it doesn't exist anymore:
```
stream.node.mount.opts.fs.stat(stream.nfd)
```

I guess there was no test for SEEK_END.

**Comment by vadimkantorov** [comment]
I could not find a stat call that takes in a node :(

**Comment by vadimkantorov** [comment]
It seems that `stream.nfd.node.usedBytes` can be used instead of this call

**Comment by vadimkantorov** [comment]
Then everything works. If this qualifies as a PR, I could send it :) but for a real thing, a test would be needed

**Comment by vadimkantorov** [comment]
About making FS's a less daunting thing to write: making ERRNO_CODES / SEEK_END/CUR/SET codes public would help! The WASM errno codes seem very different from linux.

**Comment by sbc100** [comment]
Patches certainly welcome!   

What do you mean by "making ERRNO_CODES / SEEK_END/CUR/SET codes public"?    Do you mean making them usable from user library.js files?  As far as I know they are available.

Or do you mean making them available in header files?  In which case they already are, you should be able to find them all in the headers that are part of the libc that we ship.  Have a look in `system/include/libc` (e.g. 
`system/include/libc/fcntl.h`) and also in `system/include/wasi/api.h`.

**Comment by vadimkantorov** [comment]
> What do you mean by "making ERRNO_CODES / SEEK_END/CUR/SET codes public"

I meant exporting them on Module object or some other place like that. maybe these constants are already exportable with EXTRA_...._METHODS?

**Comment by vadimkantorov** [comment]
Now that I'm rerunning the code, `stream.nfd.node.usedBytes` isn't defined anymore. Maybe I did sth wrong last time. But the problem with non-existent fstat is definitely it.

**Comment by vadimkantorov** [comment]
`stream.node.node_ops.getattr(stream.node).size` seems the stable one

**Comment by sbc100** [comment]
> > What do you mean by "making ERRNO_CODES / SEEK_END/CUR/SET codes public"
> 
> I meant exporting them on Module object or some other place like that. maybe these constants are already exportable with EXTRA_...._METHODS?

If you are writing library code you should be able to do e.g. `{{{ cDefine('SEEK_END') }}}` if you re writing C/C++ code you should be able to use `SEEK_END` directly.   Is there some other use case this doesn't cover?

**Comment by vadimkantorov** [comment]
I created a patched PROXYFS just in client JavaScript, without plugging it in the compilation process

**Comment by curiousdannii** [comment]
> Is there some other use case this doesn't cover?

I wrote an external FS, so had no access to constants like that, but I only needed about 10 of them, so it was no big deal.

What was tricky was finding where the WASI error codes were defined.

**Comment by vadimkantorov** [comment]
Indeed, I had copied them from Chrome Console during debugging session.

**Comment by junjihashimoto** [comment]
Hi,
The test of proxyfs isn't enough, but it's here.
https://github.com/emscripten-core/emscripten/blob/13493fd0a200e809e1d0b45a1dd77ed89683a93c/tests/test_other.py#L2505-L2738

**Comment by robertaboukhalil** [comment]
@kripken I think this issue can be closed now that #14706 was merged 😀

**Comment by kripken** [comment]
Yes, thanks @robertaboukhalil !
